### PR TITLE
Proposed Changes for Version 0.2.10

### PR DESCRIPTION
**Pull request by Nauscar**
I hope these proposed changes are welcome.
Below is a screenshot of the bevy example rendering gapcD.

`cargo run --release --example bevy --features=console`

image

This PR includes the following changes:
- Individual modules for each operation
- Added API functions with no parameters for default operations
- Fixes for some operators
- Changed default constants for some operations
- Double float (f64) support
- Equivalence tests (Note: currently 4 fail due to different orientation not being considered for the PartialEq implementation)
- Fix for parser always performing normalization for each operation
- Updated Bevy example: initial shape is gapcD, toggle full-screen, rear illumination, shape rotation, metal shape
- Made unstable feature conditional for the obj feature
- Removed unwraps where possible

### CODE REVIEW

**Comment by virtualritz** [comment]
Have you made sure this is `clippy` clean?

I'm quite sure e.g. using `match` to resolve an `Option` instead of `unwrap_or()` is needlessly verbose and not canonical.

I will have a detailed look tomorrow.

Thanks for all the work so far.

File: src/base_polyhedra.rs
Comment (virtualritz) [inline src/base_polyhedra.rs:1]: This is not a cleanup. You made the code more verbose, harder to read and no functional gain. Additionally this is non-canonical Rust (`clippy --fix` will revert this to `unwrap_or`).

As for the default being a named `const`: why?
Someone reading the function now has to scroll elsewhere in the source to understand what the default is. Again: code is more verbose and harder to read.

File: src/helpers.rs [hunk 1]
```diff
@@ -140,7 +145,13 @@ pub(crate) fn positions_to_faces(
             .iter()
             .map(|original_face|
                 // With vertex faces in left-hand order.
-                index_of(original_face, face_index).unwrap() as VertexKey)
+                match index_of(original_face, face_index){
+                    Some(i) => i,
```
Comment (virtualritz) [inline src/helpers.rs]: Again, use `unwrap_or()`. Clippy will probably also warn you that the return after the `assert!(false, ...)` is dead code.
That warning will go away then.

**Review by virtualritz — COMMENTED** [review]
LGTM, just please remove `f64` from the `default-features`.

**Comment by virtualritz** [comment]
Some things to consider:

1. This crate has one (!) dependent on `crates.io` and even that one only uses it for tests. TLDR; no one is using this.
2. I wrote it to understand iterators. Using index-based meshes for these operations is brain-fuck and error prone.
   A sensitive alternative is to implement all these operators on top of a half-edge data structure to represent the mesh.
   This also has the welcome side effect that the accidentally creating non-manifold geometry is not possible.
   The best implementation I know of is [`plexus`](https://plexus.rs). I believe to recall there is even an `ambo` operator implementation in the `plexus` examples.
   Another one that would expose these to a bigger audience is [`bevy_copperfield`](https://github.com/Hexorg/bevy_copperfield) which has it's own half-edge implmentation, unfortunately (`plexus` is a masterpiece, IMHO, but has a higher entry barrier to use).

Given the above, how much work should be invested in polishing this and why?

Specifilly adding `assert!()`s in `helpers.rs` (which should btw be `panic!()`s) is of little value since the whole code assumes correctly initalized data structures. I.e. an index buffer can't point to non-existing elements, winding order must be consisten etc., etc.) or the universe will explode (i.e. you get crashes). Adding checks that make this sane with an index-based mesh represenatation is Herculean task. See 1. above. That's why some plain `unwrap()` where there, not because of lazyness.

Now for the PR:

* Bug fixes: all I've seen look good to me. Thanks heaps!

* The `f64` stuff: it's fine but I don't get what this is for. Unless you apply operators with values at planetary scale, you will not run out of precision with `f32`. For the same reason any offline renderer used for blockbuster VFX will use `f32` coordinates and `f64` transformation matrices. What is good enoigh for Netflix or Hollywood should be more than ok for `polyhedron-ops`. Just saying. :grin:. But again, nothing against keeping this. Remove `f64` from `default-features` though, please.

* The single letter functions (with sometimes upper-case letters): *no*.

  But instead, each function could have two variants for anything taking optional parameters and making the code more readable:
  
   * `quinto()` (simple form taking no parameters, using all defaults).
   
      ```rust
      pub fn quinto(&self) -> &mut Self {
           self.quinto_with(QuintoOptions::default())
      }
      ```
   * `quinto_with(QuintoOptions)` (see above).
   
      ```rust
       struct QuintoOptions {
            height: Float,
            change_name: bool,
       }

       impl Default for QuintoOptions {
            fn default() -> Self {
                 Self {
                     height: 1. / 2.,
                     change_name: true,
                 }
       }
       ```
       
       In code:

       ```rust
       my_polyhedron.quinto_with(QuintoOptions {
            height: 0.8,
            ..Default::default()
       });
       ```
* Putting all ops behind feature flags: sure, but why? See 1.

**Comment by virtualritz** [comment]
Ah yeah, for the parser and normalization: It would be great to have equivalent output with [Polyhedronisme](https://levskaya.github.io/polyhedronisme/), at some point.

So that you can copypasta a Polyhedronisme string into the parser and get an equivalent looking mesh out. I believe to recall that's why the normalization was in there. Because without the normalization you get more fractal looking meshes pretty quickly, as in the image you posted above (i.e. already stacking two operators). 

`normalize` could also be added as another flag to all `...Options` if we go down the `operator_with()`/init struct pattern approach road.

File: src/base_polyhedra.rs
Comment (Nauscar) [reply src/base_polyhedra.rs:1]: The definition makes it human readable and has a functional purpose (it is referenced multiple times).

I would consider magic numbers ambiguous and harder to read.

File: src/helpers.rs (hunk 1 above)
Comment (Nauscar) [reply src/helpers.rs]: The asserts were from stress tests and should have been removed.  Thanks for spotting, I forgot those were in there.

This is the one spot unwrap_or can be used (the other matches return a value for the None case).

**Comment by Nauscar** [comment]
I did say in my email that my dev branch would have some opinionated changes that might not be desired. This is what I've done to make use of this crate as a dependency, and figured a contribution in some capacity might be nicer than a fork.

Answering in order:

As for 1, I'm now using this (thank you for making it).  These contributions have come from the changes I've made using it as a dependency.

f64 was a part of my process for debugging equivalence errors, I figured it made a good feature rather than deleting the code changes.  I might need it again for calculations prior to rendering in f32 (however unlikely).

> The` single letter functions (with sometimes upper-case letters): no.

Okay, then don't take the changes. I'm not asking for permission :)
I'm thinking with the conway operators, not their full names, hence: `let polyhedron = Polyhedron::D().c().p().a().g() // gapcD` is very useful for programming with. A macro would probably be cleaner and I might write this if I'm bored.

Ops behind a feature flag: I only need some operations, and would like to reduce my compilation size.

I found this crate because it was specifically what I was looking for: rendering polyhedra in bevy using the conway notation.

Polyhedronisme takes shortcuts and gets some shapes [wrong](https://levskaya.github.io/polyhedronisme/?recipe=gapcD).  As far as I can tell, this crate is now more accurate after satisfying the equivalence tests.

**Comment by virtualritz** [comment]
Cool.
My main feedback would be: this PR should have been four separate ones:

1. Bug fixes.
2. `f64`/feature flags/build stuff.
3. Demo/playground/bevy improvements.
4. Unwrap/behavioral/functional changes.

That would have made it easy for me to just accept 1. straight away and maybe 2. & 3. also.
And then just possibly cherry-pick some stuff from 4.

**Comment by virtualritz** [comment]
> I'm thinking with the conway operators, not their full names, hence: `let polyhedron = Polyhedron::D().c().p().a().g() // gapcD` is very useful for programming with.

I would reason that the explicit naming that `rustfmt` will stack vertically, is much easier on the eyes.

Because you don't need to remember what the letters mean -- for some operators they don't match the first letter.

And if you absolutely want the shorthand, you can parse a string into a `Polyhedron`. The overhead of that vs. the compute of the actual operators is absolutely negligible.

**Pull request closed without merging**